### PR TITLE
Support Fetch Cilium Observe data to monitoring Cilium Service network traffic

### PR DESCRIPTION
**Pull request by mrproliu**
In this PR, I mainly implemented the following features:
1. Combining data collection using [Peer](https://github.com/cilium/cilium/blob/main/api/v1/peer/peer_grpc.pb.go#L38) and [Flow](https://github.com/cilium/cilium/blob/main/api/v1/observer/observer_grpc.pb.go#L41) services provided by Cilium itself and converting it to SkyWalking data, finally displaying it in UI.
2. Added a `labelCount` function in the OAL engine for creating count statistics based on tags.
3. Rewrite FieldsHelper to support ALS as well as Cilium.

NOTE: Following https://github.com/cilium/cilium/issues/32175, we found that some flow data would be missing some metadata, so we cannot get both side metrics. Then, we decide to ignore the server-side data and convert the client-side data to the server-side. 

### All Cilium Services Overview

image
image
image

### Service Overview

image
image
image
image

### Service Relation Overview

image
image

### Service Instance Overview

image
image

### Service Instance Relaiton Overview

image

### Endpoint Overview

image
image

- [x] If it's UI related, attach the screenshots below.
- [x] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Part of #11697.
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/docs/en/changes/changes.md).

### CODE REVIEW

**Comment by wu-sheng** [comment]
UI submodule seems including more changes. Could you update the changelogs of UI side?

**Comment by wu-sheng** [comment]
How about e2e of Cilium?

**Comment by wu-sheng** [comment]
Cilium integration docs should be added in the marketplace, https://skywalking.apache.org/docs/main/next/en/setup/backend/backend-k8s-monitoring/.
And the dashboard on the UI should point to it.

File: oap-server/server-library/library-util/src/main/java/org/apache/skywalking/oap/server/library/util/FieldsHelper.java [hunk 1]
```diff
@@ -28,51 +38,43 @@
...
-
-import lombok.RequiredArgsConstructor;
-import lombok.experimental.Delegate;
-import lombok.extern.slf4j.Slf4j;
-
 @Slf4j
-public enum FieldsHelper {
-    SINGLETON;
+public class FieldsHelper {
+    private static final Map<Class<?>, FieldsHelper> HELPER_MAP = new ConcurrentHashMap<>();
```
Comment (wu-sheng) [inline oap-server/server-library/library-util/src/main/java/org/apache/skywalking/oap/server/library/util/FieldsHelper.java:49]: Could you explain this? It seems to you expand the scope of this class to cover more field processes?

File: oap-server/server-fetcher-plugin/fetcher-proto/src/main/proto/cilium/hubble/relay/relay.proto [hunk 2]
```diff
@@ -0,0 +1,42 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
```
Comment (wu-sheng) [inline oap-server/server-fetcher-plugin/fetcher-proto/src/main/proto/cilium/hubble/relay/relay.proto:2]: I think we should update the LICENSE of source codes, as we copied these from a new project.

**Comment by mrproliu** [comment]
> UI submodule seems including more changes. Could you update the changelogs of UI side?

Updated.

**Comment by mrproliu** [comment]
> How about e2e of Cilium?

I will trying to create other PR for the E2E Testing.

File: oap-server/server-fetcher-plugin/fetcher-proto/src/main/proto/cilium/hubble/relay/relay.proto (hunk 2 above)
Comment (mrproliu) [reply oap-server/server-fetcher-plugin/fetcher-proto/src/main/proto/cilium/hubble/relay/relay.proto:2]: I have update the LICENSE file, please check.

File: oap-server/server-library/library-util/src/main/java/org/apache/skywalking/oap/server/library/util/FieldsHelper.java (hunk 1 above)
Comment (mrproliu) [reply oap-server/server-library/library-util/src/main/java/org/apache/skywalking/oap/server/library/util/FieldsHelper.java]: Updated. Yes, Before we need to use `FieldsHelper.SINGLE` for getting the instance, but it only works for the ALS message. But for now, the cilium fetcher still needs this for the inflate metadata, so I make it an instance with the target inflate class. When the user needs to get the instance of `FieldsHelper` needs to call `forClass(Class)`.

**Comment by mrproliu** [comment]
> Cilium integration docs should be added in the marketplace, https://skywalking.apache.org/docs/main/next/en/setup/backend/backend-k8s-monitoring/. And the dashboard on the UI should point to it.
> 
> The docs should include
> 
> 1. Integration architecture and data flow
> 2. API dependencies
> 3. Process logic and the brief about how we process the data
> 4. How OAP cluster works in fetching mode.

Thanks. I have update the documentation following your mentioned.

File: docs/en/setup/backend/backend-k8s-monitoring.md [hunk 3]
```diff
@@ -13,6 +13,9 @@ ways to monitor deployments on Kubernetes.
 2. Rover is a SkyWalking native eBPF agent to collect network Access Logs to support topology-aware and metrics 
    analysis. Meanwhile, due to the power of eBPF, it could profile running services written by C++, Rust, Golang, etc. 
    Read [Rover setup guide](./backend-k8s-monitoring-rover.md) for more details.
+3. If Cilium is installed in Kubernetes, use Cilium Fetcher can collect network traffic data from services.
```
Comment (wu-sheng) [inline docs/en/setup/backend/backend-k8s-monitoring.md:16]: Besides this, we should have a menu item for Cilium integration.

```suggestion
3. If Cilium is installed in Kubernetes, use Cilium Fetcher to collect network traffic data of services through Cilium Hubble APIs. 
```

Is this correct?

Comment (mrproliu) [reply docs/en/setup/backend/backend-k8s-monitoring.md]: Correct. And I have update the menu.

File: docs/en/setup/backend/backend-k8s-monitoring-cilium.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
+# Kubernetes (K8s) monitoring from Rover
+
+SkyWalking uses the Cilium Fetcher to gather traffic data between services from Cilium Hubble via the Observe API. It then leverages the [OAL System](./../../concepts-and-designs/oal.md) for metrics and entity analysis.
+
+## Data flow
+
+SkyWalking Fetch Cilium Node and Observe Data from gRPC API, analysis to generate entity and using [OAL](./../../concepts-and-designs/oal.md) to generating metrics.
```
Comment (wu-sheng) [inline docs/en/setup/backend/backend-k8s-monitoring-cilium.md:7]:
```suggestion
SkyWalking fetches Cilium Node and Observability Data from gRPC API, analysis to generate entity and using [OAL](./../../concepts-and-designs/oal.md) to generating metrics.
```

File: docs/en/setup/backend/backend-k8s-monitoring-cilium.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+1. [Peers API](https://github.com/cilium/cilium/blob/main/api/v1/peer/peer_grpc.pb.go#L33-L39): Listen the hubble node in the cluster, OAP would communicate with Hubble node to obtain Observe data.
+2. [Observe API](https://github.com/cilium/cilium/blob/main/api/v1/observer/observer_grpc.pb.go#L41): Fetch the Flow data from Hubble node.
+
+## Setup
+1. Please following the [Setup Hubble Observability documentation](https://docs.cilium.io/en/stable/gettingstarted/hubble_setup/) to setting the Hubble for provided API.
+2. Setup cilium receiver module by the following configuration.
+```yaml
+cilium-fetcher:
+    selector: ${SW_CILIUM_FETCHER:default}
```
Comment (wu-sheng) [inline docs/en/setup/backend/backend-k8s-monitoring-cilium.md:19]: This is OFF by default, right?

Comment (mrproliu) [reply docs/en/setup/backend/backend-k8s-monitoring-cilium.md:19]: Yes.

File: docs/en/setup/backend/backend-k8s-monitoring-cilium.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+SkyWalking Fetch Cilium Node and Observe Data from gRPC API, analysis to generate entity and using [OAL](./../../concepts-and-designs/oal.md) to generating metrics.
+
+## API Requirements
+
+1. [Peers API](https://github.com/cilium/cilium/blob/main/api/v1/peer/peer_grpc.pb.go#L33-L39): Listen the hubble node in the cluster, OAP would communicate with Hubble node to obtain Observe data.
+2. [Observe API](https://github.com/cilium/cilium/blob/main/api/v1/observer/observer_grpc.pb.go#L41): Fetch the Flow data from Hubble node.
+
+## Setup
+1. Please following the [Setup Hubble Observability documentation](https://docs.cilium.io/en/stable/gettingstarted/hubble_setup/) to setting the Hubble for provided API.
+2. Setup cilium receiver module by the following configuration.
```
Comment (wu-sheng) [inline docs/en/setup/backend/backend-k8s-monitoring-cilium.md:16]:
```suggestion
2. Activate Cilium receiver module to set `selector=default` in the YAML or `set SW_CILIUM_FETCHER=default` through the system environment viriable.
```

File: docs/en/setup/backend/backend-k8s-monitoring-cilium.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        peerHost: ${SW_CILIUM_FETCHER_PEER_HOST:hubble-peer.kube-system.svc.cluster.local}
+        peerPort: ${SW_CILIUM_FETCHER_PEER_PORT:80}
+        fetchFailureRetrySecond: ${SW_CILIUM_FETCHER_FETCH_FAILURE_RETRY_SECOND:10}
+        sslConnection: ${SW_CILIUM_FETCHER_SSL_CONNECTION:false}
+        sslPrivateKeyFile: ${SW_CILIUM_FETCHER_PRIVATE_KEY_FILE_PATH:}
+        sslCertChainFile: ${SW_CILIUM_FETCHER_CERT_CHAIN_FILE_PATH:}
+        sslCaFile: ${SW_CILIUM_FETCHER_CA_FILE_PATH:}
+        convertClientAsServerTraffic: ${SW_CILIUM_FETCHER_CONVERT_CLIENT_AS_SERVER_TRAFFIC:true}
+```
+3. If enabled the [TLS certificate within the Hubble](https://docs.cilium.io/en/stable/gettingstarted/hubble-configuration/#tls-certificates), please update these few configuration.
```
Comment (wu-sheng) [inline docs/en/setup/backend/backend-k8s-monitoring-cilium.md:30]:
```suggestion
3. If enabled the [TLS certificate within the Hubble](https://docs.cilium.io/en/stable/gettingstarted/hubble-configuration/#tls-certificates), please update these few configurations.
```

File: docs/en/setup/backend/backend-k8s-monitoring-cilium.md [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+## Setup
+1. Please following the [Setup Hubble Observability documentation](https://docs.cilium.io/en/stable/gettingstarted/hubble_setup/) to setting the Hubble for provided API.
+2. Setup cilium receiver module by the following configuration.
+```yaml
+cilium-fetcher:
+    selector: ${SW_CILIUM_FETCHER:default}
+    default:
+        peerHost: ${SW_CILIUM_FETCHER_PEER_HOST:hubble-peer.kube-system.svc.cluster.local}
+        peerPort: ${SW_CILIUM_FETCHER_PEER_PORT:80}
```
Comment (wu-sheng) [inline docs/en/setup/backend/backend-k8s-monitoring-cilium.md:22]:
```suggestion
        # Host name and port of Hubble server
        peerHost: ${SW_CILIUM_FETCHER_PEER_HOST:hubble-peer.kube-system.svc.cluster.local}
        peerPort: ${SW_CILIUM_FETCHER_PEER_PORT:80}
```

What is the `peer`? Is there a component called Hubble server or something?

File: docs/en/setup/backend/backend-k8s-monitoring-cilium.md [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+Based on each transfer data analysis, extract the information of the 7-layer network protocol.
+
+NOTE: By default, Cilium only reports L4 metrics. If you need L7 metrics, 
+they must be explicitly specified in each service's CiliumNetworkPolicy. For details please [refer to this document](https://docs.cilium.io/en/latest/security/).
+
+#### HTTP
+
+| Name                 | Unit        | Description                                             |
+|----------------------|-------------|---------------------------------------------------------|
+| Call CPM             | Count       | HTTP Request calls per minutes.                         |
```
Comment (wu-sheng) [inline docs/en/setup/backend/backend-k8s-monitoring-cilium.md:73]:
```suggestion
|CPM             | Count       | HTTP Request calls per minutes.                         |
```
C in CPM is the calls.

File: docs/en/setup/backend/backend-k8s-monitoring-cilium.md [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+| Duration             | Nanoseconds | Total HTTP Response use duration.                       |
+| Success CPM          | Count       | Total HTTP Response success(status < 500) count.        |
+| Status 1/2/3/4/5xx   | Count       | HTTP Response status code group by 1xx/2xx/3xx/4xx/5xx. |
```
Comment (wu-sheng) [inline docs/en/setup/backend/backend-k8s-monitoring-cilium.md:76]: Is 2xx in the same graph? 2xx will be super high.

File: docs/en/setup/backend/backend-k8s-monitoring-cilium.md (hunk 8 above)
Comment (mrproliu) [reply docs/en/setup/backend/backend-k8s-monitoring-cilium.md:22]: peer is a service of the Hubble. the cilium fetcher needs getting all cilium nodes through this service. 

```shell
$kubectl get services -A
NAMESPACE      NAME                            TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)               AGE
default        details                         ClusterIP   34.118.228.1     <none>        9080/TCP              42h
default        kubernetes                      ClusterIP   34.118.224.1     <none>        443/TCP               42h
default        productpage                     ClusterIP   34.118.235.150   <none>        9080/TCP              42h
default        ratings                         ClusterIP   34.118.239.54    <none>        9080/TCP              42h
default        reviews                         ClusterIP   34.118.229.30    <none>        9080/TCP              42h
gmp-system     alertmanager                    ClusterIP   None             <none>        9093/TCP              42h
gmp-system     gmp-operator                    ClusterIP   34.118.231.151   <none>        8443/TCP,443/TCP      42h
istio-system   elasticsearch-master            ClusterIP   34.118.234.99    <none>        9200/TCP,9300/TCP     21h
istio-system   elasticsearch-master-headless   ClusterIP   None             <none>        9200/TCP,9300/TCP     21h
istio-system   skywalking-oap                  ClusterIP   34.118.228.217   <none>        11800/TCP,12800/TCP   21h
istio-system   skywalking-ui                   ClusterIP   34.118.226.129   <none>        80/TCP                21h
kube-system    default-http-backend            NodePort    34.118.233.215   <none>        80:31226/TCP          42h
kube-system    hubble-peer                     ClusterIP   34.118.232.156   <none>        443/TCP               42h
kube-system    hubble-relay                    ClusterIP   34.118.233.50    <none>        80/TCP                178m
kube-system    kube-dns                        ClusterIP   34.118.224.10    <none>        53/UDP,53/TCP         42h
kube-system    metrics-server                  ClusterIP   34.118.228.67    <none>        443/TCP               42h
```

File: docs/en/setup/backend/backend-k8s-monitoring-cilium.md (hunk 10 above)
Comment (mrproliu) [reply docs/en/setup/backend/backend-k8s-monitoring-cilium.md:76]: Yes, They are in the same graph.

File: docs/en/setup/backend/backend-k8s-monitoring-cilium.md [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+SkyWalking fetches Cilium Node and Observability Data from gRPC API, analysis to generate entity and using [OAL](./../../concepts-and-designs/oal.md) to generating metrics.
+
+## API Requirements
+
+1. [Peers API](https://github.com/cilium/cilium/blob/main/api/v1/peer/peer_grpc.pb.go#L33-L39): Listen the hubble node in the cluster, OAP would communicate with Hubble node to obtain Observe data.
+2. [Observe API](https://github.com/cilium/cilium/blob/main/api/v1/observer/observer_grpc.pb.go#L41): Fetch the Flow data from Hubble node.
+
+## Setup
+1. Please following the [Setup Hubble Observability documentation](https://docs.cilium.io/en/stable/gettingstarted/hubble_setup/) to setting the Hubble for provided API.
+2. Activate Cilium receiver module to set `selector=default` in the YAML or `set SW_CILIUM_FETCHER=default` through the system environment viriable.
```
Comment (wu-sheng) [inline docs/en/setup/backend/backend-k8s-monitoring-cilium.md:16]:
```suggestion
2. To activate Cilium receiver module, set `selector=default` in the YAML or `set SW_CILIUM_FETCHER=default` through the system environment variable.
```

File: docs/en/setup/backend/backend-k8s-monitoring-cilium.md (hunk 10 above)
Comment (wu-sheng) [reply docs/en/setup/backend/backend-k8s-monitoring-cilium.md]: I think we should list all traffic in one graph and not-2xx in one graph. This should be more clear. Traffic and Anormal Traffic

**Comment by wu-sheng** [comment]
Two questions
1. Is Hubble peer always deployed as one pod? Because I only see one hostname and one port.
2. How does OAP cluster work with Hubble peer? How many pods of OAP are going to fetch and process metrics from Hubble peer?

File: docs/en/setup/backend/backend-k8s-monitoring-cilium.md (hunk 10 above)
Comment (mrproliu) [reply docs/en/setup/backend/backend-k8s-monitoring-cilium.md]: like this?
image

**Comment by mrproliu** [comment]
> 1. Is Hubble peer always deployed as one pod? Because I only see one hostname and one port.

Hubble peer service is correlated with the cilium node, not a only pod. And Kubernetes will load balance with wich pod should be listen. 

> 2. How does OAP cluster work with Hubble peer? How many pods of OAP are going to fetch and process metrics from Hubble peer?

I have write a selection about this: [**Load Balance for Cilium Fetcher with OAP cluster**](https://github.com/apache/skywalking/pull/12393/files#diff-c709949b64ba7a07a8c5fe555a5760148f81eb7c9c429f167a0b57570aa952bcR97-R103). All the OAP nodes will listen bubble peer service for getting nodes. And each OAP node will calucate how many cilium node should be monitoring.

**Review by wu-sheng — APPROVED** [review]

File: oap-server/server-fetcher-plugin/fetcher-proto/pom.xml [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+            </goals>
+            <configuration>
+              <sources>
+                <source>${fbs.generated.sources}</source>
+              </sources>
+            </configuration>
+          </execution>
+        </executions>
+      </plugin>
+    </plugins>
```
Comment (kezhenxu94) [inline oap-server/server-fetcher-plugin/fetcher-proto/pom.xml:162]: Do we really need this bunch of plugins here?

File: oap-server/server-fetcher-plugin/cilium-fetcher-plugin/src/main/java/org/apache/skywalking/oap/server/fetcher/cilium/nodes/CiliumNode.java [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+ */
+@RequiredArgsConstructor
+@Getter
+@Slf4j
+@EqualsAndHashCode(of = {
+    "address"
+})
+@ToString(of = {
+    "address",
+    "connected"
```
Comment (kezhenxu94) [inline oap-server/server-fetcher-plugin/cilium-fetcher-plugin/src/main/java/org/apache/skywalking/oap/server/fetcher/cilium/nodes/CiliumNode.java:45]: This field doesn’t even exist in the class

File: oap-server/server-fetcher-plugin/fetcher-proto/pom.xml (hunk 12 above)
Comment (mrproliu) [reply oap-server/server-fetcher-plugin/fetcher-proto/pom.xml:162]: The fbs and flatbuffers can be deleted.

File: oap-server/server-fetcher-plugin/cilium-fetcher-plugin/src/main/java/org/apache/skywalking/oap/server/fetcher/cilium/nodes/CiliumNode.java (hunk 13 above)
Comment (mrproliu) [reply oap-server/server-fetcher-plugin/cilium-fetcher-plugin/src/main/java/org/apache/skywalking/oap/server/fetcher/cilium/nodes/CiliumNode.java:45]: `connected` can be deleted, I think I just forget to delete it when field changes.

File: oap-server/server-fetcher-plugin/fetcher-proto/src/main/proto/cilium/hubble/flow/flow.proto
Comment (kezhenxu94) [inline oap-server/server-fetcher-plugin/fetcher-proto/src/main/proto/cilium/hubble/flow/flow.proto:1]: I think these file need to be added to the LICENSE file https://github.com/apache/skywalking/blob/87b73cb7f7a7dc2e8e51daf58a04cb3c747ca91a/LICENSE#L215-L226

**Pull request merged**